### PR TITLE
Sockets - Hana

### PR DESCRIPTION
**Pull request by hanalways**
# Media Ranker
Congratulations! You're submitting your assignment!

## Comprehension Questions
Question | Answer
---      | ---
Describe a custom model method you wrote. | In the work model I wrote an instance method `entries` with an argument `category` where it would return the array of records in the category sorted by highest vote count.
Describe how you approached testing that model method. What edge cases did you come up with? | If publication year for a work was after the current year.
What are `session` and `flash`? What is the difference between them? | `flash` only hangs around for 1 request cycle and `session` lasts until the browser closed (the `session` ends), or until it’s set to `nil` manually.
What was one thing that you gained more clarity on through this assignment? | Testing controllers and how routes are directing request/responses.
What is the Heroku URL of your deployed application? | hanaworks.heroku-app,.com (NOT WORKING)
Do you have any recommendations on how we could improve this project for the next cohort? | Oof, more in class project time. Encouragement to directly apply the things in lecture to the project at hand? I felt myself completely zoning out in so many lectures and having a hard time staying focused. I know that it was stage 2, but some of the things I learned in class I wanted to implement immediately but got backburnered (`session` and `flash`) and then some lessons felt not necessarily immediately applicable to the project (`application helpers`). The whole lesson on prettifying time passed felt really disjointed from the project.

### CODE REVIEW

File: app/views/works/index.html.erb [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+    <% Work.entries(category).each do |work| %>
+    <tbody>
+      <tr>
+        <td><%= Work.get_num_votes(work) %> votes</td>
+        <td><%= link_to work.title, work_path(work) %></td>
+        <td><%= work.creator %></td>
+        <td><%= work.publication_year %></td>
+        <td>
+          <%= button_to "Upvote", new_vote_path(work), method: :post %>
+        </td>
```
Comment (droberts-sea) [inline app/views/works/index.html.erb:24]: This button doesn't seem to work. You might want `create_vote_path(@work)`

File: config/routes.rb [hunk 2]
```diff
@@ -0,0 +1,13 @@
+Rails.application.routes.draw do
+  root "homepage#index"
+  get "/", to: "homepages#index"
+  
+  resources :works, :users, :votes
```
Comment (droberts-sea) [inline config/routes.rb:5]: I don't think you need all 7 RESTful routes for `users` and `votes`.

File: app/controllers/votes_controller.rb [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+    puts "USER_ID IS: #{@vote.user_id}."
+
+    if @vote.save
+      flash[:success] = "Successfully voted for work with title \"#{Work.find_by(id: params[:work_id]).title}\""
+      redirect_to works_path
+
+    else
+      flash[:error] = "A problem occured"
+
+      redirect_back(fallback_location: root_path)
```
Comment (droberts-sea) [inline app/controllers/votes_controller.rb:18]: It would be good to give a little more information here about what went wrong, both for the user and for you as the engineer working on the product.

File: app/controllers/works_controller.rb [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+
+    @work.destroy
+    flash[:success] = "Successfully destroyed work \"#{@work.title}\""
+    redirect_to works_path
+  end
+
+  def vote
+    if session[:user_id]
+      voter_id = session[:user_id]
+      vote = Vote.new(user_id: voter_id, work_id: @work.id)
```
Comment (droberts-sea) [inline app/controllers/works_controller.rb:75]: You have controller actions for both `votes#create` and `works#vote`. Probably only one of these is needed (I think `votes#create` is the one used by the app).

File: app/controllers/votes_controller.rb [hunk 5]
```diff
@@ -0,0 +1,21 @@
+class VotesController < ApplicationController
+  def create 
+    @vote = Vote.new(
+      user_id: 1,
+      work_id: params[:work_id]
```
Comment (droberts-sea) [inline app/controllers/votes_controller.rb:5]: Instead of setting `user_id` to 1 here, you should set it to `session[:user_id]`.

As is, every vote always comes from user 1 instead of whoever is logged in.

File: app/models/vote.rb [hunk 6]
```diff
@@ -0,0 +1,10 @@
+class Vote < ApplicationRecord
+  belongs_to :user
+  belongs_to :work
+
+  validates :work_id, uniqueness: { scope: :user_id, :message => "User has already voted for this work" }
```
Comment (droberts-sea) [inline app/models/vote.rb:5]: Good work getting this tricky `uniqueness` scope figured out!

File: app/models/work.rb [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+
+  def self.entries(category)
+    return Work.where(category: category).sort_by { |work|
+      work.votes.count
+    }.reverse
+  end
+
+  def self.top_ten(category)
+    return entries(category)[0..9]
+  end
```
Comment (droberts-sea) [inline app/models/work.rb:20]: I love the way the methods in this file build on each other, each providing a small bit of functionality. This is an excellent example of functional decomposition.

File: app/views/homepage/index.html.erb [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+  Media Spotlight:
+  <%= @spotlight.title %> by <%= @spotlight.creator %>
+</section>
+
+<div class="container">
+  <div class="row">
+    <% Work.categories.each do |category| %>
+    <div class="col">
+      <p class="column head">Top <%= category.capitalize %>s</p>
+      <% Work.top_ten(category).each do |work| %>
```
Comment (droberts-sea) [inline app/views/homepage/index.html.erb:11]: Great work keeping this view DRY.

File: test/controllers/homepage_controller_test.rb [hunk 9]
```diff
@@ -0,0 +1,7 @@
+require "test_helper"
+
+describe HomepageController do
+  # it "must be a real test" do
+  #   flunk "Need real tests"
+  # end
```
Comment (droberts-sea) [inline test/controllers/homepage_controller_test.rb:6]: Looks like this was forgotten. Aside from the nominal case, there's an interesting edge case when there are no works, since the spotlight section might fail to render.

File: test/controllers/users_controller_test.rb [hunk 10]
```diff
@@ -0,0 +1,64 @@
...
+      }.wont_change "User.count"
+
+      expect(session[:user_id]).must_equal user.id
+      expect(flash[:success]).must_equal "Successfully logged in as an existing user #{user.username}"
+    end
+
+    it "will create new user" do 
+      user_data = {
+        user: {
+          username: "new user",
```
Comment (droberts-sea) [inline test/controllers/users_controller_test.rb:33]: Good work getting both of the success cases here. There is also a failure case you're missing: what if the user can't be created? For example, if the username is "".

File: test/controllers/votes_controller_test.rb [hunk 11]
```diff
@@ -0,0 +1,7 @@
+require "test_helper"
+
+describe VotesController do
+  # it "must be a real test" do
+  #   flunk "Need real tests"
+  # end
```
Comment (droberts-sea) [inline test/controllers/votes_controller_test.rb:6]: There are some interesting test cases around voting that I wish you had had time to address.

**Comment by droberts-sea** [comment]
# Media Ranker
## What We're Looking For

Feature | Feedback
---     | ---
**Core Requirements** |
Git hygiene | yes
Comprehension questions | yes
**General** |
Rails fundamentals (RESTful routing, use of named paths) | some extra routes generated
Views are well-organized (DRY, use of semantic HTML, use of partials) | yes - good work!
Errors are reported to the user | sometimes - this is definitely worth working on! Doing a good job of reporting errors is not just good for your users, it will make your life as a developer much easier.
Business logic lives in the models | yes
Models are thoroughly tested, including relations, validations and any custom logic | Would love to see test cases for the `entries` and `top_ten` methods. What happens when there are no works, or no votes? What if there are more or less than 10 works?
Controllers are thoroughly tested, including the login/logout process and multi-step workflows like voting for a work | mostly - see inline
**Wave 1 - Media** |
Splash page shows the three media categories | yes
Basic CRUD operations on media are present and functional | yes - You can't delete a work that has votes! This is because you've created a link between the two tables at the database level, and deleting the work would leave behind invalid votes. You might want to check out the [dependent](http://guides.rubyonrails.org/association_basics.html#dependent) argument to `has_many`.
**Wave 2 - Users and Votes** |
Users can log in and log out | yes
The ID of the current user is stored in the session | yes
A user cannot vote for the same media more than once | yes
All media lists are ordered by vote count | yes
Splash page contains a media spotlight | yes, but it displays a random work instead of the top-voted one
**Wave 3 - Users and Votes** |
Media pages contain lists of voting users | yes
Individual user pages and the user list are present | User list doesn't work. Individual user pages are there but vote lists are empty.
**Optional - Styling** |
Bootstrap is used appropriately | yes
Look and feel is similar to the original | this gets pretty close
**Overall** | Good work overall. While this submission has some rough edges, the core functionality matches the demo site and it is clear to me that the learning goals for this project were met. Keep up the hard work!
